_posts/2023-11-16-quarkus-meets-langchain4j.adoc

```diff
@@ -0,0 +1,236 @@
+---
+layout: post
+title: 'When Quarkus meets LangChain4j'
+date: 2023-11-15
+tags: AI langchain
+synopsis: 'Learn about the new quarkus-langchain4j extension to integrate LLMs in Quarkus applications.'
+author: cescoffier
+---
+:imagesdir: /assets/images/posts/llm
+
+Large language models (LLMs) are reshaping the world of software, altering the way we interact with users and develop business logic.
+
+Populatized by https://openai.com/[OpenAI]'s https://chat.openai.com/[ChatGPT], LLMs are now available in many flavors and sizes. The https://huggingface.co/models[Hugging-Face] platform references hundreds of them, and major tech companies like Facebook, Google, Microsoft, and Amazon are also providing their own models.
+
+LLMs are not a new concept. They have been around for a while, but they were not as powerful as they are today. The Quarkus team have been thinking about what it would mean to integrate LLMs in Quarkus applications. The talk https://www.youtube.com/watch?app=desktop&v=BD1MSLbs9KE[Java Meets AI] from Lize Raes at Devoxx 2023 has been a great source of inspiration.
+
+Since, the Quarkus team, in collaboration with the Langchain4j team, has been working on an extension to integrate LLMs in Quarkus applications. This extension is based on the https://github.com/langchain4j[Langchain4j library], which provides a common API to interact with LLMs. The Langchain4j project is a Java re-implementation of the famous https://www.langchain.com/[langchain] library.
+
+In this blog post, we will see how to use the https://docs.quarkiverse.io/quarkus-langchain4j/dev/index.html[quarkus-langchain4j] extension to integrate LLMs in Quarkus applications. This extension is an exploration to understand how LLMs can be used in Quarkus applications.
+
+== Overview
+
+First, let's have a look at the big picture. When integrating an LLM into a Quarkus application, you need to describe what you want the AI to do. Unlike traditional code, you are going to explain the behavior of the AI using natural language. Of course, there are a few techniques to tame the AI, but we will explore that later.
+
+Strictly relying on the LLM's knowledge might not be enough. Thus, the Quarkus Langchain4j extension provides two mechanisms to extend AI knowledge:
+
+- 1) Tools - a tool lets the LLM execute actions in your application. For instance, you can use a tool to send an email, call a REST endpoint, or execute a database query. The LLM decides when to use the tool, the method parameters, and what to do with the result.
+- 2) Document stores - LLMs are not good at remembering things. In addition, their context has a size limit. Thus, the extension provides a way to store and retrieve information from document stores. Before calling the LLM, the extension can ask for relevant documents in a document store and attach them to the context. The LLM can then use this data to make a decision. For instance, you can load spreadsheet data, reports, or data from a database.
+
```

Comment:
this is all true but I feel like we are repeating the same approach I feel others communicating about AI - introducing it using new terms and approaches when 90% of this is truly just simple stuff and that lots of existing java and quarkus functionality fits well into it.

Reply:
Not sure about this. I read the doc the other day (which presents the same things) and I found it extremely informative and useful to put names on concepts.

Reply:
note: I wasn't referring to Document stores specifically - but its not a big deal after rereading it - it is using more "normal" words than others i've seen so +1.